lib/seahorn/Houdini.cc

```diff
@@ -0,0 +1,292 @@
+			bvars.push_back(bvar);
+			bvar_count ++;
+		}
+	}
+
+	//What if there's no bvar?
+	if(bvar_count == 0)
+	{
+		cand = mk<TRUE>(fdecl->efac());
+	}
+	// if there is only one bvar, create a int constant and make an lt expr
+	else if(bvar_count == 1)
+	{
+		Expr zero = mkTerm<mpz_class> (0, fdecl->efac());
+		cand = mk<LT>(bvars[0], zero);
+	}
+	// if there are more than two bvars, make an lt expr
+	else if(bvar_count == 2)
+	{
+		Expr lt1 = mk<LT>(bvars[0], bvars[1]);
+		Expr lt2 = mk<LT>(bvars[1], bvars[0]);
+		bins.push_back(lt1);
+		bins.push_back(lt2);
+
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+	else // bvar_count > 2
+	{
+		for(int j=0; j<bvars.size()-1; j++)
+		{
+			Expr lt = mk<LT>(bvars[j], bvars[j+1]);
+			bins.push_back(lt);
+		}
+		cand = mknary<AND>(bins.begin(), bins.end());
+	}
+
+	return cand;
+  }
+
+  #define SAT true
+  #define UNSAT false
+
+  std::map<Expr, Expr> Houdini::currentCandidates;	//a map from predicates to candidates
+
+  /*
+   * Build the map from predicates to candidates
+   */
+  void Houdini::guessCandidate(HornClauseDB &db)
+  {
+	  for(Expr rel : db.getRelations())
+	  {
+		  if(bind::isFdecl(rel))
+		  {
+			  Expr cand = relToCand(rel);
+			  currentCandidates.insert(std::pair<Expr, Expr>(rel, cand));
+		  }
+	  }
+  }
+
+  /*
+   * Main loop of Houdini algorithm
+   */
+  void Houdini::runHoudini(HornClauseDB &db, HornifyModule &hm)
+  {
+	  guessCandidate(db);
+
+	  HornClauseDB::RuleVector workList;
+	  workList.insert(workList.end(), db.getRules().begin(), db.getRules().end());
+
+	  while(!workList.empty())
+	  {
+		  HornRule r = workList.front();
+		  errs() << "RULE HEAD: " << *(r.head()) << "\n";
+		  errs() << "RULE BODY: " << *(r.body()) << "\n";
+		  while (validateRule(r, db, hm) == SAT)
+		  {
+			  workList = addUsedRulesBackToWorkList(db, workList, r.head());
+			  weakenRuleHeadCand(r);
+		  }
+		  workList.erase(workList.begin());
+	  }
+  }
+
+  /*
+   * Given a rule, check if it's validate
+   */
+  bool Houdini::validateRule(HornRule r, HornClauseDB &db, HornifyModule &hm)
+  {
+	  Expr ruleHead_cand_app = applyActualArgsToCand(r.head());
+	  Expr neg_ruleHead_cand_app = mk<NEG>(ruleHead_cand_app);
+
+	  Expr ruleBody = r.body();
+	  ExprMap body_map;
+	  ExprVector body_pred_apps;
+	  get_all_pred_apps(ruleBody, db, std::back_inserter(body_pred_apps));
+
+	  for(ExprVector::iterator itr = body_pred_apps.begin(); itr != body_pred_apps.end(); ++itr)
+	  {
+		  body_map.insert(std::pair<Expr, Expr>(*itr, applyActualArgsToCand(*itr)));
+	  }
+
+	  Expr body_cand_app = replace(ruleBody, body_map);
+
+	  Expr whole_cand = mk<AND>(neg_ruleHead_cand_app, body_cand_app);
+
+	  errs() << "WHOLE CANDIDATE: " << *whole_cand << "\n";
+
+	  ZSolver<EZ3> solver(hm.getZContext ());
+	  solver.assertExpr(whole_cand);
+	  solver.toSmtLib(errs());
+	  bool isSat = solver.solve();
```

Comment:
if the result is neither true nor false you can still use the model to try
to throw out a lemma. If nothing works, remove a lemma arbitrarily. In the
worst case, the invariant is weakened to true and the algorithm stops.
On Wed, Jun 29, 2016 at 00:22 Chenguang-Zhu <notifications@github.com>
wrote:

> In lib/seahorn/Houdini.cc
> <https://github.com/seahorn/seahorn/pull/33#discussion_r68883313>:
>
> > +
> > +	  for(ExprVector::iterator itr = body_pred_apps.begin(); itr != body_pred_apps.end(); ++itr)
> > +	  {
> > +		  body_map.insert(std::pair<
>
> Expr, Expr>(*itr, applyActualArgsToCand(*itr)));
> > +	  }
> > +
> > +	  Expr body_cand_app = replace(ruleBody, body_map);
> > +
> > +	  Expr whole_cand = mk<AND>(neg_ruleHead_cand_app, body_cand_app);
> > +
> > +	  errs() << "WHOLE CANDIDATE: " << *whole_cand << "\n";
> > +
> > +	  ZSolver<EZ3> solver(hm.getZContext ());
> > +	  solver.assertExpr(whole_cand);
> > +	  solver.toSmtLib(errs());
> > +	  bool isSat = solver.solve();
>
> If solver returns indeterminate for a rule, then how to deal with it? I
> think maybe we should also try to weaken it until solver can say UNSAT?
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/seahorn/seahorn/pull/33/files/cce7a302670f7c9e1a2207c15b3c5d1ac93ec881#r68883313>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFzN5xb4p6U_AMeP-eMtFY2dN_7tNoufks5qQfMjgaJpZM4I7V-j>
> .
>